modules/core/src/main/scala/org/tessellation/infrastructure/trust/storage/TrustStorage.scala

```diff
@@ -4,49 +4,120 @@ import cats.Monad
+      .of[TrustStore](TrustStore(trustInfo, OrdinalTrustMap.empty, OrdinalTrustMap.initial))
+      .map(make(_))
 
-  def make[F[_]: Monad](trust: Ref[F, Map[PeerId, TrustInfo]]): TrustStorage[F] =
+  def make[F[_]: Monad: Ref.Make](
+    trustUpdates: Option[PeerObservationAdjustmentUpdateBatch]
+  ): F[TrustStorage[F]] =
+    trustUpdates.fold(make)(batch => make.flatTap(_.updateTrust(batch)))
+
+  def make[F[_]: Monad](trustStoreRef: Ref[F, TrustStore]): TrustStorage[F] =
     new TrustStorage[F] {
 
       def updateTrust(trustUpdates: PeerObservationAdjustmentUpdateBatch): F[Unit] =
-        trust.update { t =>
-          t ++ trustUpdates.updates.map { trustUpdate =>
-            trustUpdate.id -> t.getOrElse(trustUpdate.id, TrustInfo()).copy(trustLabel = trustUpdate.trust.value.some)
+        trustStoreRef.update { store =>
+          val t = store.trust
+          val updated = t.trust ++ trustUpdates.updates.map { trustUpdate =>
+            trustUpdate.id -> t.trust
+              .getOrElse(trustUpdate.id, TrustInfo())
+              .focus(_.trustLabel)
+              .replace(trustUpdate.trust.value.some)
           }.toMap
+          store.focus(_.trust.trust).replace(updated)
+        }
+
+      def updatePredictedTrust(trustUpdates: Map[PeerId, Double]): F[Unit] =
+        trustStoreRef.update { store =>
+          val t = store.trust
+          val updated = t.trust ++ trustUpdates.map {
+            case (k, v) =>
+              k -> t.trust.getOrElse(k, TrustInfo()).focus(_.predictedTrust).replace(v.some)
+          }
+          store.focus(_.trust.trust).replace(updated)
         }
 
-      def getTrust: F[Map[PeerId, TrustInfo]] = trust.get
+      def getTrust: F[TrustMap] = trustStoreRef.get.map(_.trust)
+
+      def getCurrentOrdinalTrust: F[OrdinalTrustMap] = trustStoreRef.get.map(_.currentOrdinalTrust)
+
+      def getNextOrdinalTrust: F[OrdinalTrustMap] = trustStoreRef.get.map(_.nextOrdinalTrust)
+
+      def getPublicTrust: F[PublicTrust] = getTrust.map(_.toPublicTrust)
 
-      def getPublicTrust: F[PublicTrust] = getTrust.map { trustValue =>
-        PublicTrust(trustValue.mapFilter(_.publicTrust))
+      def updatePeerPublicTrustInfo(peerId: PeerId, publicTrust: PublicTrust): F[Unit] = trustStoreRef.update { store =>
+        val t = store.trust
+        val updated = t.peerLabels.value + (peerId -> publicTrust)
+        store.focus(_.trust.peerLabels).replace(PublicTrustMap(updated))
       }
 
-      def updatePeerPublicTrustInfo(peerId: PeerId, publicTrust: PublicTrust): F[Unit] = trust.update { t =>
-        t.updatedWith(peerId) { trustInfo =>
-          trustInfo
-            .orElse(Some(TrustInfo()))
-            .map(_.copy(peerLabels = publicTrust.labels))
+      def updateNext(ordinal: SnapshotOrdinal): F[Option[SnapshotOrdinalPublicTrust]] =
+        trustStoreRef.modify { store =>
+          val next = store.nextOrdinalTrust
+
+          if (ordinal >= next.ordinal && next.trust == TrustMap.empty) {
```

Comment:
should be eq checking

Reply:
The `SnapshotOrdinal`s are compared with `Order` and `eq` is used to compare the `TrustMap`s.